### PR TITLE
docs: Next.js Best practices example with Prisma Client extensions

### LINKED ISSUE #5103: NextJS Best Practices example with Client Extensions
### Discussed in https://github.com/prisma/prisma/discussions/20321

<div type='discussions-op-text'>

<sup>Originally posted by **kroucher** July 20, 2023</sup>
Hoping for a better understanding of how to use Client Extensions using the instantiated client from the [NextJS Best Practices guide](https://www.prisma.io/docs/guides/other/troubleshooting-orm/help-articles/nextjs-prisma-client-dev-practices).

Currently, the type of the exported client is being overridden with the following:
```ts
const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined
}
```
so when the client is extended with
```ts
export const prisma = globalForPrisma.prisma ?? new PrismaClient().$extends({
  result: {
    user: {
      fullName: {
        needs: { firstName: true, lastName: true },
        compute(user) {
          return `${user.firstName} ${user.lastName}`
        },
      },
    },
  },
})
```
Due to the exported variable being overridden when not in `production` (with `.. as { prisma: PrismaClient | undefined }`), Typescript still thinks `prisma` is just `PrismaClient`, not the extended `DynamicClientExtensionThis<Prisma.TypeMap<Args & {..` and has no knowledge of the new field.

If the Client Extension is added like this, Typescript is happy, but does not include the prevention of multiple clients being instantiated:

```ts
const prismaClient =
  globalForPrisma.prisma ||
  new PrismaClient({
    log:
      process.env.NODE_ENV === "development"
        ? ["query", "error", "warn"]
        : ["error"],
  });

const prisma = prismaClient.$extends({
  result: {
    computerConfig: {
      item: {
        compute() {
          return "Hello world!";
        },
      },
    },
  },
});

export { prisma };
```</div>

### PR DESCRIPTION
**Pull request by ruheni**
Closes #5103

### CODE REVIEW

File: content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx [hunk 1]
```diff
@@ -23,15 +23,42 @@ The solution in this case is to instantiate a single instance `PrismaClient` and
...
+  prisma: PrismaClientSingleton | undefined
 }
 
-export const prisma = globalForPrisma.prisma ?? new PrismaClient()
+export const prisma = globalForPrisma.prisma ?? getExtendedClient()
 
 if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma
 ```
 
+You can extend the Prisma Client using a Prisma Client extension by appending the `$extends` client method when instantiating Prisma Client as follows:
```
Comment (ankur-arch) [inline content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx]: Shouldn't the "the" be removed as per our style [guide](https://www.prisma.io/docs/about/prisma-docs/style-guide/prisma-product-names)? 
```suggestion
You can extend Prisma Client using a Prisma Client extension by appending the `$extends` client method when instantiating Prisma Client as follows:
```

File: content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx [hunk 2]
```diff
@@ -23,15 +23,42 @@ The solution in this case is to instantiate a single instance `PrismaClient` and
...
+            return `${user.firstName} ${user.lastName}`
+          },
+        },
+      },
+    },
+  })
+}
+```
+
 After creating this file, you can now import this `PrismaClient` instance anywhere in your Next.js `pages` as follows:
```
Comment (ankur-arch) [inline content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx]:
```suggestion
After creating this file, you can now import the extended `PrismaClient` instance anywhere in your Next.js `pages` as follows:

```

File: content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx [hunk 3]
```diff
@@ -23,16 +23,43 @@ The solution in this case is to instantiate a single instance `PrismaClient` and
...
+
+type PrismaClientSingleton = ReturnType<typeof prismaClientSingleton>
+
 const globalForPrisma = globalThis as unknown as {
-  prisma: PrismaClient | undefined
+  prisma: PrismaClientSingleton | undefined
 }
 
-export const prisma = globalForPrisma.prisma ?? new PrismaClient()
+export const prisma = globalForPrisma.prisma ?? getExtendedClient()
```
Comment (jharrell) [inline content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx]: where is `getExtendedClient` coming from?

**Review by jharrell — CHANGES_REQUESTED** [review]

File: content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx (hunk 3 above)
Comment (ankur-arch) [inline content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx]: Correction:

```suggestion
export const prisma = globalForPrisma.prisma ?? prismaClientSingleton()
```

Comment (ruheni) [inline content/300-guides/500-other/880-troubleshooting-orm/100-help-articles/400-nextjs-prisma-client-dev-practices.mdx]:
```suggestion
export const prisma = globalForPrisma.prisma ?? prismaClientSingleton()
```

**Review by jharrell — APPROVED** [review]

**Pull request merged**